Specification improvement and bug fix to sign_request

rickhanlonii opened the pull request:
Per the [OAuth Body Hash specifications](http://oauth.googlecode.com/svn/spec/ext/body_hash/1.0/oauth-bodyhash.html) (3.2), if there is no body entity, then the hash should be done over the empty string.  
  
Also, when `httplib2` handles a redirect, the follow is performed with `None` as the request body. This causes a TypeError to be thrown here when attempting to hash the body.  
  
This commit prevents the TypeError, while improving the specification alignment. 
"Win, win, win." -Michael Scott

-----
Should fix issue #112
Should close pull #113  
  
Many thanks to @holm!

joestump commented:
Is this similar to the fix or different from #163?

rickhanlonii commented:
It is similar to the fix in #163 but more correct. #163 skips the hash if the request does not have a body. My change properly hashes the empty string when there's no body, as specified in [section 3.2](http://oauth.googlecode.com/svn/spec/ext/body_hash/1.0/oauth-bodyhash.html).

I also noticed you asked for a regression test in #163, I'm happy to add that here.

joestump commented:
@rickhanlonii You answered my other question in that PR as well. Please add a regression test. @jaitaiwan I'm :+1: after the test has been added. Can you take a peak?

rickhanlonii commented:
Added a test. Before my change the test I added would fail with:

```bash
Error
Traceback (most recent call last):
  File "/Library/Python/2.7/site-packages/mock.py", line 1190, in patched
    return func(*args, **keywargs)
  File "/Users/euler/dev/python-oauth2/tests/test_oauth.py", line 1322, in test_url_with_query_string_body_none
    client.request(uri, 'GET', body=None)
  File "/Users/euler/dev/python-oauth2/oauth2/__init__.py", line 668, in request
    req.sign_request(self.method, self.consumer, self.token)
  File "/Users/euler/dev/python-oauth2/oauth2/__init__.py", line 499, in sign_request
    self['oauth_body_hash'] = base64.b64encode(sha(self.body).digest())
TypeError: must be string or buffer, not None
```

joestump commented:
MMmmmm ... regression tests. :+1: cc @jaitaiwan

joestump commented:
@jaitaiwan aaaaannnnnnnnnd Travis has arisen from the dead. PRs are building and tests are passing in 2.6 and 2.7.

jaitaiwan commented:
Fantastic. LGTM

jaitaiwan commented:
@rickhanlonii can you do me a favour and get this rebased off master?

rickhanlonii commented:
Sure will :ok_hand:

betterdiff commented on oauth2/__init__.py:
```diff
@@ -485,8 +485,14 @@ def sign_request(self, signature_method, consumer, token):
 
         if not self.is_form_encoded:
-            # according to
+            # according to 
```
* W291 trailing whitespace

betterdiff commented on oauth2/__init__.py:
```diff
@@ -485,8 +485,14 @@ def sign_request(self, signature_method, consumer, token):
-            # according to
+            # according to 
             # http://oauth.googlecode.com/svn/spec/ext/body_hash/1.0/oauth-bodyhash.html
+            # section 3.2 "If the request does not have an entity body, the hash should 
```
* E501 line too long (87 > 79 characters)
* W291 trailing whitespace

betterdiff commented on oauth2/__init__.py:
```diff
@@ -485,8 +485,14 @@ def sign_request(self, signature_method, consumer, token):
+            # be taken over the empty string."
+            if self.body is None:
+                self.body = ""
+            
```
* W293 blank line contains whitespace

betterdiff commented on tests/test_oauth.py:
```diff
@@ -1290,6 +1290,38 @@ def mockrequest(cl, ur, **kw):
+            self.failUnlessEqual(len(expected), len(actual))
+            actual = dict(actual)
+            for key, value in expected:
+                if key not in ('oauth_signature', 'oauth_nonce', 'oauth_timestamp'):
```
* E501 line too long (84 > 79 characters)

betterdiff commented on tests/test_oauth.py:
```diff
@@ -1290,6 +1290,38 @@ def mockrequest(cl, ur, **kw):
+    def test_url_with_query_string_body_none(self, mockHttpRequest):
+        uri = 'http://example.com/foo/bar/?show=thundercats&character=snarf'
+        client = oauth.Client(self.consumer, None)
+        random_result = random.randint(1,100)
```
* E231 missing whitespace after ','

betterdiff commented on tests/test_oauth.py:
```diff
@@ -1290,6 +1290,38 @@ def mockrequest(cl, ur, **kw):
+
+        def mockrequest(cl, ur, **kw):
+            self.failUnless(cl is client)
+            self.failUnlessEqual(frozenset(kw.keys()), frozenset(['method', 'body', 'redirections', 'connection_type', 'headers']))
```
* E501 line too long (131 > 79 characters)

betterdiff commented on tests/test_oauth.py:
```diff
@@ -1290,6 +1290,38 @@ def mockrequest(cl, ur, **kw):
+            self.failUnlessEqual(kw['body'], None)
+            self.failUnlessEqual(kw['connection_type'], None)
+            self.failUnlessEqual(kw['method'], 'GET')
+            self.failUnlessEqual(kw['redirections'], httplib2.DEFAULT_MAX_REDIRECTS)
```
* E501 line too long (84 > 79 characters)

betterdiff commented on tests/test_oauth.py:
```diff
@@ -1290,6 +1290,38 @@ def mockrequest(cl, ur, **kw):
+            self.failUnless(isinstance(kw['headers'], dict))
+
+            req = oauth.Request.from_consumer_and_token(self.consumer, None,
+                                                        http_method='GET', http_url=uri, parameters={})
```
* E501 line too long (103 > 79 characters)

betterdiff commented on tests/test_oauth.py:
```diff
@@ -1290,6 +1290,38 @@ def mockrequest(cl, ur, **kw):
+
+            req = oauth.Request.from_consumer_and_token(self.consumer, None,
+                                                        http_method='GET', http_url=uri, parameters={})
+            req.sign_request(oauth.SignatureMethod_HMAC_SHA1(), self.consumer, None)
```
* E501 line too long (84 > 79 characters)
